Source/DafnyLanguageServer/Language/CachingParser.cs

```diff
@@ -26,13 +27,18 @@ public class CachingParser : ProgramParser {
       telemetryPublisher, programName, "parsing");
   }
 
-  protected override DfyParseResult ParseFile(DafnyOptions options, Func<TextReader> getReader,
+  protected override DfyParseResult ParseFile(DafnyOptions options, bool parseAsDooFile, Func<TextReader> getReader,
     Uri uri, CancellationToken cancellationToken) {
     using var reader = getReader();
-    var (newReader, hash) = ComputeHashFromReader(uri, reader, HashAlgorithm.Create("SHA256")!);
+    
+    // Add NUL delimiter to avoid collisions (otherwise hash("A" + "BC") == hash("AB" + "C"))
+    var uriBytes = Encoding.UTF8.GetBytes(uri + "\0");
+    var startingBytes = new[] {Convert.ToByte(parseAsDooFile) }.Concat(uriBytes).ToArray();
```

Comment:
Why was this necessary? Shouldn't we have different URIs for a .dfy and a .doo file, even if they contain identical content?

Reply:
True, but at the parser level it's not guaranteed that the URI matches the `parseAsDooFile` bool. I could match on the URI and leave out the `parseAsDooFile` parameter altogether, but I'd prefer not to rely on the URI any more at this point.